packages/nimble-components/src/menu-button/index.ts

```diff
@@ -162,16 +162,52 @@ export class MenuButton extends FoundationElement implements ButtonPattern {
     public menuKeyDownHandler(e: KeyboardEvent): boolean {
         switch (e.key) {
             case keyEscape:
-                this.open = false;
+                this.setOpen(false);
                 this.toggleButton!.focus();
                 return false;
             default:
                 return true;
         }
     }
 
+    private setOpen(newValue: boolean): void {
+        if (this.open === newValue) {
+            return;
+        }
+
+        const eventDetail: MenuButtonBeforeToggleEventDetail = {
+            oldState: this.open,
+            newState: newValue
+        };
+        this.$emit('beforetoggle', eventDetail);
+
+        this.open = newValue;
+    }
+
     private get menu(): HTMLElement | undefined {
-        return this.slottedMenus?.length ? this.slottedMenus[0] : undefined;
+        // Get the menu that is slotted within the menu-button, taking into account
+        // that it may be nested within multiple 'slot' elements, such as when used
+        // within a table.
+        if (!this.slottedMenus?.length) {
+            return undefined;
+        }
+
+        let currentItem = this.slottedMenus[0];
+        while (currentItem) {
+            if (currentItem.getAttribute('role') === 'menu') {
+                return currentItem;
+            }
+
+            if (currentItem?.nodeName === 'SLOT') {
+                currentItem = (
+                    currentItem as HTMLSlotElement
+                ).assignedNodes()[0] as HTMLElement;
```

Comment:
if we can't avoid the `as` cast then we should try and represent it better, maybe `as HTMLElement | undefined`?

I quickly scanned the tests and don't think I saw them but do we have tests for edge cases like the most nested slot doesn't have children, has child text nodes, or has child non-menu elements? (are those useful tests?)

Reply:
I changed this code slightly to use `instanceof`, which allowed me to get rid of the `as` cast.

I also updated the tests to include a cases where a menu wasn't provided (both in the nested slot case and the standard case). I'm only running some tests in those cases because a lot of the focus behavior doesn't apply to not having a menu. I'm not sure how useful the tests are because the menu button isn't even functional in those cases since it doesn't have a menu, but it at least ensures we don't throw an exception if the timing works out that the menu doesn't exist yet when the button first opens.